### ISSUE TITLE
Prevent Data Leakage During Preprocessing in day7_ex.py

### ISSUE THREAD

**Issue by md-abu-kayser** [EimanTahir027/Model-Tuning-and-Optimization#47]
## Describe the bug

The preprocessing workflow in `code_files/7-Optimization Project - Building and Tuning a Final Model/day7_ex.py` fits preprocessing statistics on the complete dataset before the train/test split.

Specifically, `StandardScaler` is fitted on all numerical samples before `train_test_split()`:

```python
scaler = StandardScaler()
numerical_features = ['tenure', 'MonthlyCharges', 'TotalCharges']
df[numerical_features] = scaler.fit_transform(df[numerical_features])
```

The missing-value imputation also calculates the `TotalCharges` median from the complete dataset before the split:

```python
df.fillna({'TotalCharges': df['TotalCharges'].median()}, inplace=True)
```

As a result, information from samples that later become part of the test set is incorporated into the preprocessing statistics before model evaluation.

This introduces **data leakage** into the evaluation workflow.

## To Reproduce

1. Open `code_files/7-Optimization Project - Building and Tuning a Final Model/day7_ex.py`.
2. Locate the `TotalCharges` missing-value handling and numerical feature scaling sections.
3. Observe that `df['TotalCharges'].median()` is calculated before `train_test_split()`.
4. Observe that `StandardScaler.fit_transform()` is applied to the complete dataset before `train_test_split()`.
5. The dataset is subsequently split into training and test sets:

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
```

6. The resulting test set therefore contains samples whose information has already influenced the preprocessing statistics.

## Expected behavior

Preprocessing statistics should be learned exclusively from the training data.

The expected workflow is:

1. Split the dataset into training and test sets.
2. Fit missing-value imputation statistics using the training data only.
3. Fit `StandardScaler` using the training data only.
4. Apply the fitted transformations to the test data without refitting.
5. During cross-validation, preprocessing should be fitted independently within each training fold.

This prevents information from the evaluation data from influencing the training or preprocessing process.

## Screenshots

Not applicable. This is a data-processing and model-evaluation issue rather than a visual/UI issue.

## Desktop (please complete the following information)

* OS: N/A
* Browser: N/A
* Version: N/A

## Smartphone (please complete the following information)

* Device: N/A
* OS: N/A
* Browser: N/A
* Version: N/A

## Additional context

The issue affects the reliability of the model evaluation rather than causing a runtime exception.

The current workflow can produce evaluation metrics that are influenced by information from the test set. This makes the reported test performance and subsequent model assessment less reliable.

A scikit-learn `Pipeline` and, where appropriate, `ColumnTransformer` can be used to keep preprocessing and model training together and ensure that preprocessing transformations are fitted only on the relevant training data during cross-validation.

### Affected file

`code_files/7-Optimization Project-Building and Tuning a Final Model/day7_ex.py`
